Add component handlers to nosedrum for handling message component interactions

Awlexus opened the pull request.

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+  handled by a module that implements the `Nosedrum.ComponentInteraction`-
+  behaviour or any pid with the `register_components/2`-callback. 
+
+  When a pid is given to the `register_components/2`-callback, it will receive 
+  a message of the form `{:message_component_interaction, t:Nostrum.Struct.Interaction.t()}`.
```
When will it receive the message? We should probably specify that it will receive it once the the component interaction is received by the gateway. Or is it actually sent that directly when calling it?

Also, I think it would be preferable to move this message to a type, so the linking of the Interaction struct works properly. Not sure there's another way to do it. Or does this link to the type these days?

Awlexus replied:
It does not work this way. Thank you for pointing that out!

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+  a message of the form `{:message_component_interaction, t:Nostrum.Struct.Interaction.t()}`.
+
+  ## Handle incomming interactions
+  Use the `handle_message_component_interaction/1`-callback to handle 
```
```suggestion
  Use the `handle_message_component_interaction/1` callback to handle 
```

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+
+  ## Handle incomming interactions
+  Use the `handle_message_component_interaction/1`-callback to handle 
+  incomming interactions and route them to the correct module/pid. The 
```
```suggestion
  incomming interactions and route them to the correct module or pid. The 
```

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+  Use the `handle_message_component_interaction/1`-callback to handle 
+  incomming interactions and route them to the correct module/pid. The 
+  recommended place to handle these interactions would be when handling
+  the `:INTERACTION_CREATE`-event in the consumer. The following example
```
```suggestion
  the `:INTERACTION_CREATE` event in the consumer. The following example
```

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+  incomming interactions and route them to the correct module/pid. The 
+  recommended place to handle these interactions would be when handling
+  the `:INTERACTION_CREATE`-event in the consumer. The following example
+  uses the `Nosedrum.ComponentHandler.ETS`-implementation
```
```suggestion
  uses the `Nosedrum.ComponentHandler.ETS` implementation:
```

jchristgit commented on lib/nosedrum/component_handler/ets.ex:
```diff
@@ -0,0 +1,83 @@
+          Storage.followup(interaction, callback_tuple)
+        end
+
+      _ ->
```
Should we explicitly match on `[[]]` here to make sure we don't have garbage in the ETS table?

Awlexus replied:
I don't use ETS often, so I'm not sure if that's necessary. Under which circumstances would this match return `[[]]`?

jchristgit replied:
Sorry, I think I meant `[]`. I'm asking for it because the other two explicitly match on the type and if by _tiny_ chance we ever have something else inserted e.g. by a bug we can catch it here.

Awlexus replied:
Ah, that makes more sense. Good idea!

jchristgit commented on lib/nosedrum/component_handler.ex:
```diff
@@ -0,0 +1,37 @@
+    case interaction.type do
+      1 -> Nostrum.Api.create_interaction_response(interaction, %{type: 1})
+      2 -> Nosedrum.Storage.Dispatcher.handle_interaction(interaction)
+      x when x in 3..5 -> Nosedrum.ComponentHandler.ETS.handle_component_interaction(interaction)
```
Nosedrum shouldn't start any of this on its own, so maybe we should specify that `Nosedrum.ComponentHandler.ETS` should be added to the supervision tree (same for `Nosedrum.Storage.Dispatcher`?

jchristgit commented on lib/nosedrum/component_interaction.ex:
```diff
@@ -0,0 +1,7 @@
+defmodule Nosedrum.ComponentInteraction do
+  @doc """
+  Handle message component interactions. Behaves the same way as `Nosedrum.ApplicationCommand.commmand/1`-callback
```
```suggestion
  Handle message component interactions.
  
  Behaves the same way as the `Nosedrum.ApplicationCommand.commmand/1` callback.
```

jchristgit reviewed (commented):
This looks great!

I have some minor style comments. Apart from that I'm very happy with this!

Awlexus commented:
I rewrote the documentation to be hopefully more understandable. 

I also extended the component handler to require an arbitrary term to be given as a third argument, which will then be passed on to the module or pid. I believe this should eliminate the need for most use-cases to start a process just for storing some state. 

I also updated ex_doc, because there was a new documentation feature that I wanted to include. On that note I should mention that I had remove the line  [`application: [],`](https://github.com/jchristgit/nosedrum/blob/master/mix.exs#L64) from mix.exs, because otherwise mix.docs would fail for me on elixir 1.16. I'm not sure if it's still a requirement for older elixir versions, so I left it in there for now

jchristgit reviewed (approved):
Thank you very much!

jchristgit commented:
Thanks!
